scripts/mempool.js

```diff
@@ -1,343 +1,242 @@
-import { getNetwork } from './network.js';
-import { getStakingBalance } from './global.js';
-import { Database } from './database.js';
 import { getEventEmitter } from './event_bus.js';
-import Multimap from 'multimap';
-import { wallet } from './wallet.js';
-import { cChainParams } from './chain_params.js';
-import { Account } from './accounts.js';
 import { Transaction, COutpoint, UTXO } from './transaction.js';
 
-export const UTXO_WALLET_STATE = {
-    NOT_MINE: 0, // Don't have the key to spend this utxo
-    SPENDABLE: 1, // Have the key to spend this (P2PKH) utxo
-    SPENDABLE_COLD: 2, // Have the key to spend this (P2CS) utxo
-    COLD_RECEIVED: 4, // Have the staking key of this (P2CS) utxo
-    SPENDABLE_TOTAL: 1 | 2,
-    IMMATURE: 8, // Coinbase/ coinstake that it's not mature (and hence spendable) yet
-    LOCKED: 16, // Coins in the LOCK set
-};
+export const OutpointState = {
+    OURS: 1 << 0, // This outpoint is ours
 
-/**
- * A historical transaction
- */
-export class HistoricalTx {
-    /**
-     * @param {HistoricalTxType} type - The type of transaction.
-     * @param {string} id - The transaction ID.
-     * @param {Array<string>} receivers - The list of 'output addresses'.
-     * @param {boolean} shieldedOutputs - If this transaction contains Shield outputs.
-     * @param {number} time - The block time of the transaction.
-     * @param {number} blockHeight - The block height of the transaction.
-     * @param {number} amount - The amount transacted, in coins.
-     */
-    constructor(
-        type,
-        id,
-        receivers,
-        shieldedOutputs,
-        time,
-        blockHeight,
-        amount
-    ) {
-        this.type = type;
-        this.id = id;
-        this.receivers = receivers;
-        this.shieldedOutputs = shieldedOutputs;
-        this.time = time;
-        this.blockHeight = blockHeight;
-        this.amount = amount;
-    }
-}
+    P2PKH: 1 << 1, // This is a P2PKH outpoint
+    P2CS: 1 << 2, // This is a P2CS outpoint
 
-/**
- * A historical transaction type.
- * @enum {number}
- */
-export const HistoricalTxType = {
-    UNKNOWN: 0,
-    STAKE: 1,
-    DELEGATION: 2,
-    UNDELEGATION: 3,
-    RECEIVED: 4,
-    SENT: 5,
+    SPENT: 1 << 3, // This outpoint has been spent
+    IMMATURE: 1 << 4, // Coinbase/coinstake that it's not mature (hence not spendable) yet
+    LOCKED: 1 << 5, // Coins in the LOCK set
 };
 
-/** A Mempool instance, stores and handles UTXO data for the wallet */
 export class Mempool {
+    /** @type{Map<string, number>} */
+    #outpointStatus = new Map();
```

Comment:
This is very compact but slow, by taking in account benchmarks we should decide if it's better having this or the  3 smaller structures:
`multimap` for spent Outpoints
`set` for unspent Outpoints
`set` for locked Outpoints
2 more maps but would be much faster

Reply:
According to benchmarks, the bottleneck is generating new addresses to check if they're ours or not, so I'm marking this as resolved